src/main/java/org/ohdsi/webapi/shiro/filters/TeamProjectBasedAuthorizingFilter.java

```diff
@@ -0,0 +1,246 @@
+          // set to only the one requested/found in the request parameters (following lines below):
+          String login = this.authorizer.getCurrentUser().getLogin();
+          boolean foundValidTeamProject = extractAndValidateTeamProjectRoleAndUpdateUserIfNecessary(this, login, request, response);
+          if (!foundValidTeamProject) {
+            return false;
+          }
+        }
+
+      } catch (Exception e) {
+        WebUtils.toHttp(response).setHeader("x-auth-error", e.getMessage());
+        throw new Exception(e);
+      }
+
+    return true;
+  }
+
+  /**
+   * Tries to extract "team project" role found in the request.
+   * If necessary, assigns the newly found "team project" role to the user,
+   * after validating with a Gen3 authorization service if the user
+   * has been granted access to this team.
+   * 
+   * @param self: reference to "this". Just a workaround to allow the method to be
+   * globally synchronized by making it static.
+   * @param login: the user login
+   * @param request: the request passing through this filter
+   * @param response: response object that can be used to write error status and
+   * error message if needed.
+   * 
+   * @return: Returns true if the "team project" was found and passed validation
+   * and DB update to user roles did not fail.
+   * 
+   * @throws IOException
+   * @throws Exception
+   */
+  private static boolean extractAndValidateTeamProjectRoleAndUpdateUserIfNecessary(
+      TeamProjectBasedAuthorizingFilter self,
+      String login, ServletRequest request,
+      ServletResponse response)  throws IOException, Exception {
+
+    // synchronize on login to avoid race conditions (especially on DB updates) if
+    // this filter receives parallel requests for any reason:
+    synchronized (login.intern()) {
+      // check if a teamproject parameter is found in the request:
+      String teamProjectRole = self.extractTeamProjectFromRequestParameters(request);
+      Set<String> newUserRoles = new HashSet<String>();
+      Set<String> newDefaultRoles = new HashSet<String>(self.defaultRoles);
+
+      // if found, add teamproject as a role in the newUserRoles list:
+      if (teamProjectRole != null && !teamProjectRole.trim().isEmpty()) {
+        // double check if this role has really been granted to the user:
+        if (self.checkGen3Authorization(teamProjectRole, login) == false) {
+          WebUtils.toHttp(response).sendError(HttpServletResponse.SC_FORBIDDEN,
+          "User is not authorized to access this team project's data");
+          return false;
+        }
+        // add teamproject role and related system role that
+        // enables read restrictions/permissions based read access configurations:
+        newDefaultRoles.add("read restricted Atlas Users"); // aka reserved system role 15 - TODO: since this is always added...maybe it should be one of the default roles instead...
```

Comment:
to be discussed.

Reply:
I agree with you about changing this so that default roles are configurable and we don't need this here. THe default permissions seem to be set in src/main/java/org/ohdsi/webapi/shiro/management/AtlasSecurity.java and could be configured for both team projects and shared artifact owner/reader permissions